### PR TITLE
AP_Periph: replace minimal NEOPixel/Buzzer support with full AP_Notify

### PR DESCRIPTION
**Pull request by magicrub**
replace minimal NEOPixel/Buzzer support with full AP_Notify

### CODE REVIEW

**Comment by magicrub** [comment]
This has been updated to allow low-memory systems to still use NeoPixel and buzzer without having to pull in the whole Notify+Servo library. However, there's no examples to follow in any hwdef. Should I add that to the F101?

This is the example I'm referring to
```
define HAL_PERIPH_ENABLE_BUZZER
define HAL_PERIPH_NEOPIXEL_COUNT 8
define HAL_PERIPH_NEOPIXEL_CHAN 0
```

**Comment by magicrub** [comment]
@ukmook and @davidbitton pretty please can you give this a test? Thanks!

**Comment by davidbitton** [comment]
I tested and I have a strip of 8 Neopixels on a Pixracer blinking.

**Comment by ukmook** [comment]
Tested this PR all working as expected, colours match across Flight controller (pixracer LED), F413_Node & Neopixel.

**Comment by magicrub** [comment]
whoops, I rebased it on the wrong branch instead of upstream! Fixed

File: Tools/AP_Periph/AP_Periph.h [hunk 1]
```diff
@@ -142,6 +160,19 @@ class AP_Periph_FW {
...
+#if defined(HAL_PERIPH_ENABLE_NOTIFY) || defined(HAL_PERIPH_NEOPIXEL_COUNT)
+    // notification object for LEDs, buzzers etc (parameter set to
+    // false disables external leds)
+    struct AP_Periph_Notify {
+#ifdef HAL_PERIPH_ENABLE_NOTIFY
+        AP_Notify lib;
+#endif
+        uint32_t update_ms;
+
+        void update_rainbow();
```
Comment (bugobliterator) [inline Tools/AP_Periph/AP_Periph.h:172]: what is the reason for defining this under the structure?

File: Tools/AP_Periph/Parameters.cpp [hunk 2]
```diff
@@ -140,6 +140,12 @@ const AP_Param::Info AP_Periph_FW::var_info[] = {
     GOBJECT(servo_channels, "OUT",     SRV_Channels),
 #endif
 
+#ifdef HAL_PERIPH_ENABLE_NOTIFY
+    // @Group: NTF_
+    // @Path: ../libraries/AP_Notify/AP_Notify.cpp
+    GOBJECT(notify, "NTF_",  AP_Notify),
```
Comment (bugobliterator) [inline Tools/AP_Periph/Parameters.cpp:146]: shouldn't this be notify.lib?

File: Tools/AP_Periph/can.cpp [hunk 3]
```diff
@@ -483,7 +483,11 @@ static void handle_beep_command(CanardInstance* ins, CanardRxTransfer* transfer)
     fix_float16(req.duration);
     buzzer_start_ms = AP_HAL::native_millis();
     buzzer_len_ms = req.duration*1000;
+#ifdef HAL_PERIPH_ENABLE_BUZZER
     float volume = constrain_float(periph.g.buzz_volume/100.0, 0, 1);
+#elif defined(HAL_PERIPH_ENABLE_NOTIFY)
+    float volume = constrain_float(periph.notify.lib.get_buzz_volume()/100, 0, 1);
```
Comment (bugobliterator) [inline Tools/AP_Periph/can.cpp:489]: 100.0f not 100, also in the above line 100.0 should be 100.0f, we might be forcing needless double operation.

**Review by bugobliterator — CHANGES_REQUESTED** [review]

**Review by bugobliterator — CHANGES_REQUESTED** [review]
I don't understand the need for substructure notify. I believe in the case we want to derive a class from AP_Periph_FW it would be hard to change such methods to virtual. Also it seems to stray from the path things are done under AP_Periph.

File: Tools/AP_Periph/AP_Periph.h (hunk 1 above)
Comment (magicrub) [reply Tools/AP_Periph/AP_Periph.h:172]: It's LED related. In general, we're trying to group all the features into classes for better organization.

File: Tools/AP_Periph/Parameters.cpp (hunk 2 above)
Comment (magicrub) [reply Tools/AP_Periph/Parameters.cpp:146]: oh wow, yes. I wonder how this is even working!??!?!

Comment (magicrub) [reply Tools/AP_Periph/Parameters.cpp:146]: it doesn't compile as notify.lib. Also, battery is like this too. Hmm, interesting that this is the way it likes it.

File: Tools/AP_Periph/can.cpp (hunk 3 above)
Comment (magicrub) [reply Tools/AP_Periph/can.cpp:489]: fixed

**Comment by magicrub** [comment]
@bugobliterator I went back and forth on this just now, and went ahead and removed the struct. It only held 2 things and one can easily get changed to a local static like others did.

**Comment by magicrub** [comment]
I renamed the LEDs and BUZZERs defines to be more clear to be slightly more self documenting
example: ```HAL_PERIPH_ENABLE_NCP5623_LED``` to ```HAL_PERIPH_ENABLE_NCP5623_LED_WITHOUT_NOTIFY```

File: Tools/AP_Periph/Parameters.cpp (hunk 2 above)
Comment (magicrub) [reply Tools/AP_Periph/Parameters.cpp:154]: I've removed lib so this no linger applies

File: Tools/AP_Periph/AP_Periph.h (hunk 1 above)
Comment (magicrub) [reply Tools/AP_Periph/AP_Periph.h]: removed the struct

**Review by tridge — CHANGES_REQUESTED** [review]
I just tried this on the CUAV_GPS and it breaks the rainbow pattern on startup. The CUAV GPS is a production GPS so I don't want to break it's behaviour

**Comment by magicrub** [comment]
It shouldn't. I've included a default.parm file that should be all the correct settings on boot. Have you used an older version of notify PR on that board?

**Comment by tridge** [comment]
> It shouldn't. I've included a default.parm file that should be all the correct settings on boot. Have you used an older version of notify PR on that board?

I've pushed a fix to this branch, the problem was max embedded param size was zero in hwdef.dat for CUAV_GPS

**Review by tridge — APPROVED** [review]

**Pull request merged**